File: librarymanager/librarymanager-be/src/main/java/ru/spring5/elena/librarymanager/controller/ReaderController.java
package ru.spring5.elena.librarymanager.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import ru.spring5.elena.librarymanager.api.ReaderApi;
import ru.spring5.elena.librarymanager.api.dto.BookUsing;
import ru.spring5.elena.librarymanager.api.dto.Order;
import ru.spring5.elena.librarymanager.api.dto.Reader;
import ru.spring5.elena.librarymanager.services.ReaderService;

import java.util.List;

@RequestMapping(value = "librarymanager/readers", produces = {MediaType.APPLICATION_JSON_VALUE})
@RestController
@RequiredArgsConstructor
@Slf4j
public class ReaderController implements ReaderApi {
    private final ReaderService readerService;

    @Override
	@GetMapping("/{readerid}")
    public Reader getReader(@PathVariable("readerid") Long readerId) {
        return readerService.findById(readerId);
    }
    
    @Override
	@PostMapping("/search")
    public Page<Reader> searchReaders(
            Pageable pageable
    ) {
        log.info("Listing readers");
        return readerService.findAllByPage(pageable);
    }
    
    @Override
	@PostMapping
    public Reader createReader(
            @RequestBody Reader reader
    ) {
        log.info("Creating reader");
        return readerService.create(reader);
    }

    @Override
	@PutMapping("/{readerid}")
    public Reader saveReader(
            @PathVariable("readerid") Long readerId,
            @RequestBody Reader reader
    ) {
        log.info("Updating reader");
        return readerService.save(readerId, reader);
    }
    
    @Override
	@PostMapping("/{readerid}/orders/search")
    public Page<Order> searchOrders(
            @PathVariable("readerid") Long readerId,
            Pageable pageable
    ) {
        log.info("Listing orders by reader");
        return readerService.findOrdersByReaderId(readerId,pageable);
    }
    
    @Override
	@PutMapping("/{readerid}/orders/")
    public List<Order> updateOrders(
    		@PathVariable("readerid") Long readerId,
    		@RequestParam("list") List<Long> orderIds
    		)
    {
    	log.info("Update orders of reader");
    	return readerService.updateOrders(readerId, orderIds);
    }
    
    @Override
	@PostMapping("/{readerid}/bookusings/search")
    public Page<BookUsing> searchBookUsings(
            @PathVariable("readerid") Long readerId,
            Pageable pageable
    ) {
        log.info("Listing bookUsings by reader {0}", readerId);
        return readerService.findBookUsingsByReaderId(readerId,pageable);
    }
    
    @Override
	@PutMapping("/{readerid}/bookusings/")
    public List<BookUsing> updateBookUsings(
    		@PathVariable("readerid") Long readerId,
    		@RequestParam("list") List<Long> bookUsingIds
    		)
    {
    	log.info("Update bookUsings of reader {0}",readerId );
    	return readerService.updateBookUsings(readerId, bookUsingIds);
    }
}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/controller/ReaderController.kt
package ru.elenacherev.librarymanager.controller

import org.springframework.data.domain.Pageable
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import ru.elenacherev.librarymanager.api.ReaderApi
import ru.elenacherev.librarymanager.api.dto.Reader
import ru.elenacherev.librarymanager.services.ReaderService
import java.util.*

@RequestMapping("/readers")
@RestController
class ReaderController(
    val readerService: ReaderService
) : ReaderApi {

    @GetMapping("/{readerid}")
    override fun getReader(
        @PathVariable("readerid") readerId: UUID
    ) = readerService.findById(readerId)

    @PostMapping("/search")
    override fun searchReaders(
        pageable: Pageable
    ) = readerService.findAllByPage(pageable)

    @PostMapping
    override fun createReader(
        @RequestBody reader: Reader
    ) = readerService.create(reader)

    @PutMapping("/{readerid}")
    override fun saveReader(
        @PathVariable("readerid") readerId: UUID,
        @RequestBody reader: Reader
    ) = readerService.save(readerId, reader)

    @PostMapping("/{readerid}/orders/search")
    override fun searchOrders(
        @PathVariable("readerid") readerId: UUID,
        pageable: Pageable
    ) = readerService.findOrdersByReaderId(readerId, pageable)

    @PutMapping("/{readerid}/orders/")
    override fun updateOrders(
        @PathVariable("readerid") readerId: UUID,
        @RequestParam("list") orderIds: List<UUID>
    ) = readerService.updateOrders(readerId, orderIds)

    @PostMapping("/{readerid}/bookusings/search")
    override fun searchBookUsings(
        @PathVariable("readerid") readerId: UUID,
        pageable: Pageable
    ) = readerService.findBookUsingsByReaderId(readerId, pageable)

    @PutMapping("/{readerid}/bookusings/")
    override fun updateBookUsings(
        @PathVariable("readerid") readerId: UUID,
        @RequestParam("list") bookUsingIds: List<UUID>
    ) = readerService.updateBookUsings(readerId, bookUsingIds)
}